### PR TITLE
test: bc/avs slash tests

### PR DESCRIPTION
**Pull request by ypatil12**
**Motivation:**

Tests that handle slashes by both the BC and AVS onto Native ETH. 

**Modifications:**

Added assertions for BC/AVS slashings, no full slashed included

- [x] testFuzz_avsSlash_bcSlash_checkpoint
- [x] testFuzz_bcSlash_checkpoint_avsSlash
- [x] testFuzz_avsSlash_verifyValidator_bcSlash_checkpoint
- [x] testFuzz_avsSlash_bcSlash_verifyValidator_checkpoint
- [x] testFuzz_avsSlash_bcSlash_checkpoint_verifyValidator
- [x] testFuzz_avsSlash_bcSlash_balanceIncrease_checkpoint
- [x] testFuzz_avsSlash_bcSlash_checkpoint_balanceIncrease

**Result:**

Test coverage for BC/AVS edge cases. 

Need to properly bound a lot of the AVS/BC slashing cases, but this is a start

### CODE REVIEW

**Comment by 0xClandestine** [comment]
Cannot understand why ci is failing...

**Comment by bowenli86** [comment]
image
i'm getting the same error locally

File: src/test/integration/IntegrationBase.t.sol [hunk 1]
```diff
@@ -562,6 +564,28 @@ abstract contract IntegrationBase is IntegrationDeployer, TypeImporter {
...
+
+    function assert_withdrawableSharesDecreasedByAtLeast(
+        User staker,
+        IStrategy strategy,
+        uint256 originalShares,
+        uint256 expectedDecreases,
+        string memory err
+    ) internal view {
+        return assert_withdrawableSharesDecreasedByAtLeast(staker, strategy.toArray(), originalShares.toArrayU256(), expectedDecreases.toArrayU256(), err);
+    }
```
Comment (nadir-akhtar) [inline src/test/integration/IntegrationBase.t.sol:588]: Curious why you had the below function call into the above function and transform each parameter into an array, rather than have the function taking in arrays call into the other function

File: src/test/integration/IntegrationBase.t.sol [hunk 2]
```diff
@@ -1282,6 +1306,127 @@ abstract contract IntegrationBase is IntegrationDeployer, TypeImporter {
         }
     }
 
+    /*******************************************************************************
+                    SNAPSHOT ASSERTIONS: BEACON CHAIN AND AVS SLASHING
+    *******************************************************************************/
+    //TODO: handle all of these bounds properly
```
Comment (nadir-akhtar) [inline src/test/integration/IntegrationBase.t.sol:1312]: Assuming tighter bounds caused tests to fail?

File: src/test/integration/IntegrationBase.t.sol [hunk 3]
```diff
@@ -1282,6 +1306,127 @@ abstract contract IntegrationBase is IntegrationDeployer, TypeImporter {
         }
     }
 
+    /*******************************************************************************
+                    SNAPSHOT ASSERTIONS: BEACON CHAIN AND AVS SLASHING
+    *******************************************************************************/
+    //TODO: handle all of these bounds properly 
+
+    /// @dev Same as above, but when a BC slash occurs before an AVS slash
```
Comment (nadir-akhtar) [inline src/test/integration/IntegrationBase.t.sol:1314]: I think this comment might be outdated since there's nothing above that seems relevant

File: src/test/integration/IntegrationBase.t.sol [hunk 4]
```diff
@@ -1282,6 +1306,127 @@ abstract contract IntegrationBase is IntegrationDeployer, TypeImporter {
...
+    /// @dev There is additional rounding error when a BC and AVS slash occur together
+    function assert_Snap_StakerWithdrawableShares_AfterBCSlash_AVSSlash(
+        User staker,
+        AllocateParams memory allocateParams,
+        SlashingParams memory slashingParams,
+        string memory err
+    ) internal {
+        require(allocateParams.strategies.length == 1  && slashingParams.strategies.length == 1, "only beacon strategy supported");
```
Comment (nadir-akhtar) [inline src/test/integration/IntegrationBase.t.sol:1322]: Super nit: I noticed two spaces where there only needs to be one
```suggestion
        require(allocateParams.strategies.length == 1 && slashingParams.strategies.length == 1, "only beacon strategy supported");
```

File: src/test/integration/IntegrationBase.t.sol [hunk 5]
```diff
@@ -1282,6 +1306,126 @@ abstract contract IntegrationBase is IntegrationDeployer, TypeImporter {
...
+
+            uint256 slashedShares = 0;
+
+            if (slashingParams.strategies.contains(strat)) {
+                uint wadToSlash = slashingParams.wadsToSlash[slashingParams.strategies.indexOf(strat)];
+                slashedShares = prevShares[i].mulWadRoundUp(allocateParams.newMagnitudes[i].mulWadRoundUp(wadToSlash));
+            }
+
+            assertEq(prevShares[i] - slashedShares, curShares[i], err);
+        }
```
Comment (nadir-akhtar) [inline src/test/integration/IntegrationBase.t.sol:1340]: Since we know that we only have 1 strategy and that it's the beacon chain strat (based on the above `require()` statements), we can simplify this quite a bit. Specifically, we can:
* Remove this loop (since there's only 1)
* Remove the conditional `if` (since we know they're the same)
* We can also assume the `indexOf` response (`0`) (since, again, we only have 1 specific strategy)

Figure we can simplify it to look something like this:

```suggestion
        uint curShares = _getWithdrawableShares(staker, allocateParams.strategies)[0];
        uint prevShares = _getPrevWithdrawableShares(staker, allocateParams.strategies)[0];

        uint wadToSlash = slashingParams.wadsToSlash[0];
        uint slashedShares = prevShares[i].mulWadRoundUp(allocateParams.newMagnitudes[i].mulWadRoundUp(wadToSlash));

        assertEq(prevShares - slashedShares, curShares, err);
```

File: src/test/integration/IntegrationBase.t.sol [hunk 6]
```diff
@@ -1500,6 +1644,24 @@ abstract contract IntegrationBase is IntegrationDeployer, TypeImporter {
         }
     }
 
+    /// @dev This is currently used
+    /// TODO: Figure out how to bound this better
+    function assert_Snap_Added_Staker_WithdrawableSharesAtLeast(
```
Comment (nadir-akhtar) [inline src/test/integration/IntegrationBase.t.sol:1649]: Incomplete comment, and it seems that the name doesn't match the behavior (`AtLeast` is in the name, but there's no inequality check). Also seems to be true for a couple of the other added assertions

Also recommend suffixing with `_AtLeast` for naming consistency

File: src/test/integration/IntegrationChecks.t.sol [hunk 7]
```diff
@@ -1008,4 +1007,63 @@ contract IntegrationCheckUtils is IntegrationBase {
         assert_Snap_Removed_AllocatedSet(operator, allocateParams.operatorSet, "should not have updated allocated sets");
         assert_Snap_Removed_AllocatedStrats(operator, allocateParams.operatorSet, slashParams.strategies, "should not have updated allocated strategies");
     }
+
+    /*******************************************************************************
+                                 BC/AVS SLASHING CHECKS
+    *******************************************************************************/
+
+    function check_CompleteCheckPoint_AfterAVS_BCSlash(
```
Comment (nadir-akhtar) [inline src/test/integration/IntegrationChecks.t.sol:1015]: Nit: naming consistency
```suggestion
    function check_CompleteCheckpoint_AfterAVSSlash_BCSlash(
```

File: src/test/integration/tests/DualSlashing.t.sol [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+
+    function _init() internal virtual override {
+        super._init();
+
+        // 8. Slash operator by AVS
+        slashingParams = _genSlashing_Rand(operator, operatorSet);
+        avs.slashOperator(slashingParams);
+        assert_Snap_Allocations_Slashed(slashingParams, operatorSet, true, "operator allocations should be slashed");
+        assert_Snap_Unchanged_Staker_DepositShares(staker, "staker deposit shares should be unchanged after slashing");
+        assert_Snap_StakerWithdrawableShares_AfterBCSlash_AVSSlash(staker, allocateParams, slashingParams, "staker withdrawable shares should be slashed");
```
Comment (nadir-akhtar) [inline src/test/integration/tests/DualSlashing.t.sol:96]: The use of this assertion is interesting, at least based on the name, since a BC slash has not happened yet but we're using this check

Also curious if it makes sense to use `check_Base_Slashing_State` here in addition to / instead of these specific assertions

**Review by nadir-akhtar — COMMENTED** [review]
Test logic overall looks solid -- had some comments for my own knowledge and to simplify some logic (as well as a few whitespace / typo nits) @ypatil12

File: src/test/integration/IntegrationBase.t.sol (hunk 1 above)
Comment (ypatil12) [reply src/test/integration/IntegrationBase.t.sol]: Good callout. Fixed: https://github.com/Layr-Labs/eigenlayer-contracts/pull/1153/commits/c70cbde00966351863e4cdf5048a222e51dce125

File: src/test/integration/IntegrationBase.t.sol (hunk 2 above)
Comment (ypatil12) [reply src/test/integration/IntegrationBase.t.sol:1312]: Yup, this rounding is expected though. We've made `assert_Snap_StakerWithdrawableShares_AfterAVSSlash_BCSlash` have strict equalities

File: src/test/integration/IntegrationBase.t.sol (hunk 3 above)
Comment (ypatil12) [reply src/test/integration/IntegrationBase.t.sol]: It's actually same as the above assertion, clarified: https://github.com/Layr-Labs/eigenlayer-contracts/pull/1153/commits/26d8b13b74a5669a798e247a6405ed4f7007168a

File: src/test/integration/IntegrationBase.t.sol (hunk 4 above)
Comment (ypatil12) [reply src/test/integration/IntegrationBase.t.sol]: https://github.com/Layr-Labs/eigenlayer-contracts/pull/1153/commits/df2845f6cfecc60481d08d533e0ffd05b12369c0

File: src/test/integration/IntegrationBase.t.sol (hunk 6 above)
Comment (ypatil12) [reply src/test/integration/IntegrationBase.t.sol]: https://github.com/Layr-Labs/eigenlayer-contracts/pull/1153/commits/8e319159798a6a7f9c1c190285e2221b1383f03f

File: src/test/integration/IntegrationChecks.t.sol (hunk 7 above)
Comment (ypatil12) [reply src/test/integration/IntegrationChecks.t.sol]: https://github.com/Layr-Labs/eigenlayer-contracts/pull/1153/commits/f5098ada59dc4fc0673f1698c2dad0b20a22743f

File: src/test/integration/IntegrationBase.t.sol (hunk 5 above)
Comment (ypatil12) [reply src/test/integration/IntegrationBase.t.sol]: Good suggestion, implemented: https://github.com/Layr-Labs/eigenlayer-contracts/pull/1153/commits/672d5ae35acdde5c66a61d390312e905009d714c

File: src/test/integration/tests/DualSlashing.t.sol (hunk 8 above)
Comment (ypatil12) [reply src/test/integration/tests/DualSlashing.t.sol]: Great suggestions, added `check_Base_Slashing_State` and instead used `assert_Snap_StakerWithdrawableShares_AfterSlash` for the last assertion :) https://github.com/Layr-Labs/eigenlayer-contracts/pull/1153/commits/b7621c74c68a95630874509c845467534197272c

**Review by nadir-akhtar — APPROVED** [review]
🎉
